### ISSUE TITLE
GRMustache is incompatible with Swift 5

### ISSUE THREAD

**Issue by djones6** [IBM-Swift/Kitura-MustacheTemplateEngine#28]
Our fork of GRMustache has diverged from Gwendal Roué's version over the years.  It is still in Swift 3 format, running on Swift 4.x in 3.0 compatibility mode, which means it will not build at all on Swift 5.

GRMustache itself has not been updated in 18 months, though [there are a number of PRs to address Swift 4 support and bug fixes](https://github.com/groue/GRMustache.swift/pulls).

We need to either deprecate `Kitura-MustacheTemplateEngine`, making it clear that it will not support Swift 5, or put some effort into updating our GRMustache fork to be Swift 4 compatible.

**Comment by askari01** [comment]
kitura mustache dependency install fails.
with Xcode 10.2 (swift 5)

**Comment by ianpartridge** [comment]
Yes, I'm afraid the upstream project we depend on needs some love :( Looks like it's not going to become compatible with Swift 5.

**Comment by askari01** [comment]
Correct me if I am wrong. GRMustache is no longer being maintained and it is a dependency Kitura-MacheteTemplate.

**Comment by ianpartridge** [comment]
Yes that's right. We may have to deprecate Kitura-MustacheTemplateEngine

**Comment by askari01** [comment]
do we need to deprecate it, it is being widely used.

**Comment by ianpartridge** [comment]
That's interesting to hear, I was not aware of widespread usage. Stencil seemed to be much more popular. Do you have any information about its usage?

**Comment by ianpartridge** [comment]
Confirmation that upstream support has ceased: https://twitter.com/groue/status/1082253076432605184

**Comment by ianpartridge** [comment]
Good news!  There may be a new upstream maintainer: https://github.com/groue/GRMustache.swift/pull/57#issuecomment-485765239

If there is a new release from upstream supporting Swift 5 we would be glad to support it in `Kitura-MustacheTemplateEngine`.

**Comment by pushkarnk** [comment]
I see that there've been no updates to the ObjC code in the upstream repo over the last for years. This means all of the new commits that we need to cherry-pick onto the IBM-Swift fork must be pure Swift. However, before cherry-picking the new commits I wanted to make sure the IBM-Swift fork builds with Swift 5 - its not. Working on that now.

**Comment by pushkarnk** [comment]
- I'm now able to get the IBM-Swift fork building with Swift 5. All the tests also pass (using `make test`). 
- With 5.1 the build fails because SwiftJSON has a build failure (looking into it)
- The IBM-Swift fork is 98 commits behind and 420 commits ahead of upstream GRMustache. I see that one of the biggest changes to GRMustache (https://github.com/groue/GRMustache.swift/pull/35) wasn't "cherry-picked" (if I may), possibly for some genuine reasons in the past. Absorbing that change will be a considerable effort.

**Comment by pushkarnk** [comment]
Raised https://github.com/IBM-Swift-Sunset/GRMustache.swift/pull/14

**Comment by djones6** [comment]
https://github.com/groue/GRMustache.swift/issues/66 hasn't received any comments - I'd suggest we raise an upstream PR with the work done so far (assuming it's in a good state) and let the PR process determine what else needs changing.

**Comment by adam-rocska** [comment]
@djones6 No chance this'd ever get resolved, right? 😞

**Comment by djones6** [comment]
I don't think it did, I'm afraid.  IIRC there were two parts to this:
1. Get the IBM-Swift fork of GRMustache to compile again on Swift 5 - that work was being done under: https://github.com/IBM-Swift-Sunset/GRMustache.swift/pull/14
2. Upstream the changes for Linux support - which appears to need a significant effort due to the amount the two have diverged over the years.

If we could confirm that the PR above is good + get that merged, it might at least get GRMustache functional again on Swift 5.  I'm no longer actively working on Swift, but if someone could help review that PR (test that it works - we don't have any CI on that repo) then I imagine we could push it through.

**Comment by adam-rocska** [comment]
Wow. @djones6  Thanks for the quick reply!
I would normally volunteer to get my hands wet, or at least assign someone from my side. However we're still fresh with Swift, and with IBM Kitura, so I don't think our competencies would be there just yet.

However we're really happy with the stack so far, so as soon as we get to a confident professional level, we'll sure as hell start contributing wherever we can.
